controllers/nodemaintenance_controller_test.go

```diff
@@ -2,338 +2,367 @@ package controllers
-			Expect(err).NotTo(HaveOccurred())
-			_ = AddOrRemoveTaint(r.drainer.Client, node, true)
-			taintedNode := &corev1.Node{}
-			err = k8sClient.Get(context.TODO(), client.ObjectKey{Name: "node01"}, taintedNode)
-			Expect(err).NotTo(HaveOccurred())
-			Expect(taintExist(taintedNode, "medik8s.io/drain", corev1.TaintEffectNoSchedule)).To(BeTrue())
-			Expect(taintExist(taintedNode, "node.kubernetes.io/unschedulable", corev1.TaintEffectNoSchedule)).To(BeTrue())
-			Expect(taintExist(taintedNode, "test", corev1.TaintEffectPreferNoSchedule)).To(BeTrue())
-
-			// there is a not-ready taint now as well... skip count tests
-			//Expect(len(taintedNode.Spec.Taints)).To(Equal(3))
-		})
-
-		It("should remove medik8s NoSchedule taint and keep other existing taints", func() {
-			node := &corev1.Node{}
-			err := k8sClient.Get(context.TODO(), client.ObjectKey{Name: "node01"}, node)
-			Expect(err).NotTo(HaveOccurred())
-			Expect(taintExist(node, "medik8s.io/drain", corev1.TaintEffectNoSchedule)).To(BeFalse())
-			_ = AddOrRemoveTaint(r.drainer.Client, node, true)
-			taintedNode := &corev1.Node{}
-			err = k8sClient.Get(context.TODO(), client.ObjectKey{Name: "node01"}, taintedNode)
-			Expect(err).ToNot(HaveOccurred())
-			Expect(taintExist(taintedNode, "medik8s.io/drain", corev1.TaintEffectNoSchedule)).To(BeTrue())
-			_ = AddOrRemoveTaint(r.drainer.Client, taintedNode, false)
-			unTaintedNode := &corev1.Node{}
-			err = k8sClient.Get(context.TODO(), client.ObjectKey{Name: "node01"}, unTaintedNode)
-			Expect(err).NotTo(HaveOccurred())
-			Expect(taintExist(unTaintedNode, "medik8s.io/drain", corev1.TaintEffectNoSchedule)).To(BeFalse())
-			Expect(taintExist(unTaintedNode, "test", corev1.TaintEffectPreferNoSchedule)).To(BeTrue())
-
-			//Expect(len(unTaintedNode.Spec.Taints)).To(Equal(1))
+		Context("Testing Taints", func() {
+			BeforeEach(func() {
+				Expect(k8sClient.Create(context.Background(), nodeOne)).To(Succeed())
+				DeferCleanup(k8sClient.Delete, context.Background(), nodeOne)
+			})
+			When("Adding new taint", func() {
+				It("should add new taint and keep other existing taints", func() {
+					node := &corev1.Node{}
+					Expect(k8sClient.Get(context.Background(), client.ObjectKey{Name: taintedNodeName}, node)).To(Succeed())
+					Expect(AddOrRemoveTaint(r.drainer.Client, node, true)).To(Succeed())
+					taintedNode := &corev1.Node{}
+					Expect(k8sClient.Get(context.Background(), client.ObjectKey{Name: taintedNodeName}, taintedNode)).To(Succeed())
+					Expect(isTaintExist(taintedNode, medik8sDrainTaint.Key, medik8sDrainTaint.Effect)).To(BeTrue())
+					Expect(isTaintExist(taintedNode, NodeUnschedulableTaint.Key, NodeUnschedulableTaint.Effect)).To(BeTrue())
+					Expect(isTaintExist(taintedNode, dummyTaintKey, corev1.TaintEffectPreferNoSchedule)).To(BeTrue())
+					// there is also a not-ready taint
+					Expect(len(taintedNode.Spec.Taints)).To(Equal(4))
+				})
+			})
+
+			When("Adding and then removing a taint", func() {
```

Comment:
Originally, I was trying to keep the logic of the old tests and not remove some as it could leave us with missing tests, and there is a small value in keeping the current behavior. 
In the beginning, I was more keen on going with the adding + removing direction.
The value I see here is that we separate dependency between adding and removing (not necessarily here..., but the addition could be not as part of the test) of taint/label. Another argument for adding + removing direction is that by using "steps" (once for addition and for removal) we can gain this nice separation for debugging the cause of a possible error.

Do you see this dependency separation as valuable for these functionality tests?

Reply:
> The value I see here is that we separate dependency between adding and removing

Either I don't understand what you mean, or i just don't see any value 🤔 

> we can gain this nice separation for debugging the cause of a possible error.

you always get the line number of a failed test 🤷🏼‍♂️